connectors/byoei.py

```diff
@@ -499,6 +569,8 @@ async def async_bulk(
         - sync_rules_enabled: if enabled, applies rules -- default: `False`
         - options: dict of options (from `elasticsearch.bulk` in the config file)
         """
+        if self._fetcher_task is not None or self._bulker_task is not None:
+            raise AsyncBulkRunningError("Async bulk task has already started.")
```

Comment:
It feels a bit weird - is it possible to move initialisation of `fetcher` and `bulker` to the initialiser?

The pattern of "This class's method X can be called only once but its signature shows that it can be used for different purposes" is super confusing to me.

E.g. `async_bulk` takes index name as a parameter, meaning that single ElasticServer instance can serve ingestion into multiple indices. However, if you try to do so, the error will be raised - it's counter-intuitive and really confusing to read IMO.

Reply:
>E.g. `async_bulk` takes index name as a parameter, meaning that single ElasticServer instance can serve ingestion into multiple indices.

I believe it was designed like this in the first place, and that's why there's only one instance of `ElasticServer` instantiated in `JobSchedulingService` and shared by all sync jobs:
https://github.com/elastic/connectors-python/blob/3cfabc3f0b366da7ab900042c752525f8b3ad65f/connectors/services/job_scheduling.py#L141

But actually when you call `async_bulk`, it creates a new `Fetcher` object, a new `Bulker` object and a new `MemQueue`:
https://github.com/elastic/connectors-python/blob/3cfabc3f0b366da7ab900042c752525f8b3ad65f/connectors/byoei.py#L473
https://github.com/elastic/connectors-python/blob/3cfabc3f0b366da7ab900042c752525f8b3ad65f/connectors/byoei.py#L485
https://github.com/elastic/connectors-python/blob/3cfabc3f0b366da7ab900042c752525f8b3ad65f/connectors/byoei.py#L464 

The purpose of this PR is to move this flow to the background
https://github.com/elastic/connectors-python/blob/3cfabc3f0b366da7ab900042c752525f8b3ad65f/connectors/byoei.py#L16-L19

So that `async_bulk` will return immediately after `bulker` and `fetcher` tasks are created, without waiting for it to complete. 

Therefore I want to keep the change as small as possible. I moved the instantiation of `ElasticServer` to each `SyncJobRunner` because I need the object to hold `bulker_task` and `fetcher_task`, so it won't be able to be shared by multiple sync jobs.

Ideally, I would like to make it work like what I described here: https://github.com/elastic/connectors-python/pull/528#discussion_r1143601281.